File: code/99.java
// method 1 stack
class Solution {
    public void recoverTree(TreeNode root) {
        TreeNode swap1=null, swap2=null, prev=null, cur;
        Stack<TreeNode> stack = new Stack<>();
        // Deque<TreeNode> stack = new LinkedList<>(); this line works as well.
        cur = root;
        while (cur != null || !stack.isEmpty()) {
            while (cur!= null) {
                stack.push(cur);
                cur = cur.left;
            }
            cur = stack.pop();
            if (prev != null && prev.val > cur.val) {
                if (swap1 == null) {
                    swap1 = prev;
                    swap2 = cur;
                } else {
                    swap2 = cur;
                    break;
                }
            }
            prev = cur;
            cur = cur.right;
        }
        int temp = swap1.val;
        swap1.val = swap2.val;
        swap2.val = temp;
    }
}

// method 2 recursion

class Solution {
    TreeNode val1 = null, val2 = null;
    TreeNode min = null;
    public void recoverTree(TreeNode root) {
        walk(root);
        int tmp = val1.val;
        val1.val = val2.val;
        val2.val = tmp;
    }
    private void walk(TreeNode curr){
        if(curr == null) return;
        walk(curr.left);
        if(min != null && curr.val < min.val){
            if(val1 == null){
                val1 = min;
                val2 = curr;
            }else
                val2 = curr; //override previous val2 since there are only two misplaced elems
        }
        min = curr;
        walk(curr.right);
    }
}

File: code/99.py
# Method 1 recursion
class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        def inorder(root):
            if not root:
                return
            nonlocal x, y, pre
            inorder(root.left)
            if pre and pre.val > root.val:
                y = root
                if not x:
                    x = pre
                else:
                    return
            pre = root
            inorder(root.right)

        x = y = pre = None
        inorder(root)
        x.val, y.val = y.val, x.val


# Method 2(stack)
class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        cur = root
        prev = None
        swap1 = swap2 = None
        stack = []
        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left

            cur = stack.pop()
            if prev and prev.val > cur.val:
                if not swap1:
                    swap1 = prev
                    swap2 = cur
                else:
                    swap2 = cur
                    break

            prev = cur
            cur = cur.right

        swap1.val, swap2.val = swap2.val, swap1.val
        return root
